Objective-C/CBLReadOnlyArray.h

```diff
@@ -0,0 +1,112 @@
+//
+//  CBLReadOnlyArray.h
+//  CouchbaseLite
+//
+//  Created by Pasin Suriyentrakorn on 4/12/17.
+//  Copyright © 2017 Couchbase. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import "CBLReadOnlyArrayFragment.h"
+@class CBLBlob;
+@class CBLReadOnlySubdocument;
+@class CBLReadOnlyArray;
+
+NS_ASSUME_NONNULL_BEGIN
+
+/** CBLReadOnlyArray protocol defines a set of methods for readonly accessing array data. */
+@protocol CBLReadOnlyArray <NSObject, CBLReadOnlyArrayFragment>
+
+/** Gets a number of the items in the array. */
+@property (readonly) NSUInteger count;
+
+/** Gets value at the given index as an object. The object types are CBLBlob,
+    CBLReadOnlyArray, CBLReadOnlyDictionary, NSNumber, or NSString based on the underlying
+    data type; or nil if the value is NSNull.
+    @param index    the index.
+    @result the object or nil. */
+- (nullable id) objectAtIndex: (NSUInteger)index;
+
+/** Gets value at the given index as a boolean. 
+    Returns YES if the value exists, and is either `true` or a nonzero number.
+    @param index    the index.
+    @result the boolean value. */
+- (BOOL) booleanAtIndex: (NSUInteger)index;
+
+/** Gets value at the given index as an integer.
+    Floating point values will be rounded. The value `true` is returned as 1, `false` as 0.
+    Returns 0 if the property doesn't exist or does not have a numeric value.
+    @param index    the index.
+    @result the integer value. */
+- (NSInteger) integerAtIndex: (NSUInteger)index;
+
+/** Gets value at the given index as a float.
+    Integers will be converted to float. The value `true` is returned as 1.0, `false` as 0.0.
+    Returns 0.0 if the property doesn't exist or does not have a numeric value.
+    @param index    the index.
+    @result the float value. */
+- (float) floatAtIndex: (NSUInteger)index;
+
+/** Gets value at the given index as a double.
+    Integers will be converted to double. The value `true` is returned as 1.0, `false` as 0.0.
+    Returns 0.0 if the property doesn't exist or does not have a numeric value.
+    @param index    the index.
+    @result the double value. */
+- (double) doubleAtIndex: (NSUInteger)index;
+
+/** Gets value at the given index as a string.
+    Returns nil if the property doesn't exist, or its value is not a string.
+    @param index    the index.
+    @result the NSString object or nil. */
+- (nullable NSString*) stringAtIndex: (NSUInteger)index;
+
+/** Gets value at the given index as a number.
+    Returns nil if the property doesn't exist, or its value is not a number.
+    @param index    the index.
+    @result the NSNumber object or nil. */
+- (nullable NSNumber*) numberAtIndex: (NSUInteger)index;
```

Comment:
Does Java have an equivalent class?  C# does not.  The reason that NSNumber exists is to represent differently sized numbers as one type but that reason is irrelevant for managed languages.  When porting this, I simply skipped this method.